soh/src/overlays/actors/ovl_En_Tk/z_en_tk.c

```diff
@@ -611,13 +611,21 @@ void EnTk_Dig(EnTk* this, GlobalContext* globalCtx) {
                  * Upgrade the purple rupee reward to the heart piece if this
                  * is the first grand prize dig.
                  */
-                if (!(gSaveContext.itemGetInf[1] & 0x1000)) {
-                    gSaveContext.itemGetInf[1] |= 0x1000;
+                // Initialized with bugfixed condition, overriden with original check if bugfix cvar is 0 (which is default).
+                bool condition = Flags_GetCollectible(globalCtx, 0x19) == 0;
+                if (CVar_GetS32("gGravediggingTourFix", 0) == 0) {
+                    condition = !(gSaveContext.itemGetInf[1] & 0x1000);
+                }
+                if (condition) {
                     this->currentReward = 4;
+                    gSaveContext.itemGetInf[1] |= 0x1000;
                 }
             }
```

Comment:
```suggestion
                if (!(gSaveContext.itemGetInf[1] & 0x1000)
                    || (CVar_GetS32("gGravediggingTourFix", 0) != 0 && !Flags_GetCollectible(globalCtx, 0x19))) {
                    gSaveContext.itemGetInf[1] |= 0x1000;
                    this->currentReward = 4;
                }
            }
```

This should make the logic look a little cleaner. if the itemGetInf flag isn't set, it should always enter the if statement (assuming nothing else sets that flag), otherwise we check if the fix should be applied and that the collectible flag hasn't been set yet. Not familiar with how `CVar_GetS32` is used, but I imagine the `!= 0` isn't necessary unless it's needed for style reasons.

Reply:
You are correct that the `!= 0` probably isn't required. I didn't feel like experimenting with whether or not C automatically saw zero as false and nonzero as true so I just went explicit with the boolean expression. I have since double checked that it does so I could change that.

As for the larger change here, wouldn't this make it so that if someone triggered the glitch with the box unchecked by getting the heart piece and leaving, they would then be unable to get the Heart Piece by checking the box and trying again? The old flag still gets set to preserve the vanilla behavior, and the `OR` clause in your change would then return `true` with that flag set whether the `CVar` for the fix is `1` or `0`. My current code already covers this case, by not checking the vanilla flag if the fix is enabled, but if it would be better for readability I could take your check and add another `&&` for the CVar being 0. So the full condition check would be

```c
if (!(CVar_GetS32("gGravediggingTourFix", 0)) && !(gSaveContext.itemGetInf[1] & 0x1000)
    || CVar_GetS32("gGravediggingTourFix", 0 && !(Flags_GetCollectible(globalCtx, 0x19)))) {
```

I have doubts if that is really more readable though. That's a pretty lengthy boolean expression and has multiple calls to `CVar_GetS32("gGravediggingTourFix, 0)` whereas my current code only has one.

Reply:
> As for the larger change here, wouldn't this make it so that if someone triggered the glitch with the box unchecked by getting the heart piece and leaving, they would then be unable to get the Heart Piece by checking the box and trying again?

it shouldn't. the simplified pseudocode here is this:

    if (!itemGetInf) OR (fixEnabled AND !collectibleFlag) {
        set itemGetInf true
        set reward to heart container
    }

The optimization here is that if the itemGetInf flag is false, it is safe to assume that the collectibleFlag is also false regardless of whether the fix should be applied or not. This is because itemGetInf flips to true when the heart container is spawned for the first time, and if the heart container hasn't ever spawned, then collectibleFlag can't be true. The OR here simply introduces an extra alternative condition for spawning the heart container. 

Toggling the option should not lead to duplicates or making the heart permanently skippable; this is mainly because with the code written this way, the itemGetInf state and collectibleFlag state is always set the same way, independent of the state of the fix flag.

Reply:
OH I see, you're right, I think I misinterpreted the check on the old flag as it being collected returning true rather than not collected. That is a good point that I didn't see right away and the last point you made is also good.